src/ssh.c

```diff
@@ -1191,41 +1191,59 @@ int wolfSSH_SetChannelType(WOLFSSH* ssh, byte type, byte* name, word32 nameSz)
     return WS_SUCCESS;
 }
 
-int wolfSSH_SetUsername(WOLFSSH* ssh, const char* username)
+
+int wolfSSH_SetUsernameRaw(WOLFSSH* ssh,
+        const byte* username, word32 usernameSz)
 {
-    char* value = NULL;
-    word32 valueSz = 0;
+    char* newUsername = NULL;
     int ret = WS_SUCCESS;
 
-    if (ssh == NULL || ssh->handshake == NULL ||
-        ssh->ctx->side == WOLFSSH_ENDPOINT_SERVER ||
-        username == NULL) {
-
+    if (ssh == NULL)
+        ret = WS_BAD_ARGUMENT;
+    if (username == NULL || usernameSz == 0)
         ret = WS_BAD_ARGUMENT;
-    }
 
     if (ret == WS_SUCCESS) {
-        valueSz = (word32)WSTRLEN(username);
-        if (valueSz > 0)
-            value = (char*)WMALLOC(valueSz + 1, ssh->ctx->heap, DYNTYPE_STRING);
-        if (value == NULL)
+        newUsername = (char*)WMALLOC(usernameSz + 1,
+                ssh->ctx->heap, DYNTYPE_STRING);
+        if (newUsername == NULL)
             ret = WS_MEMORY_E;
+        else {
+            WMEMCPY(newUsername, username, usernameSz);
+            newUsername[usernameSz] = 0;
+            if (ssh->userName != NULL)
+                WFREE(ssh->userName, ssh->ctx->heap, DYNTYPE_STRING);
+            ssh->userName = newUsername;
```

Comment:
case here of a leak if called twice in a row without free'ing the previous pointer

Reply:
also should we make sure a null terminator is set here? hard setting the last value to '\0' for safety